pkg/datapath/loader/netkit.go

```diff
@@ -0,0 +1,187 @@
+//
+// Returns nil if the update was successful. Returns an error wrapping
+// [os.ErrNotExist] if the link is defunct or missing.
+func updateNetkit(prog *ebpf.Program, progName, bpffsDir string) error {
+	// Attempt to open and update an existing link.
+	pin := filepath.Join(bpffsDir, progName)
+	err := bpf.UpdateLink(pin, prog)
+	switch {
+	// Link exists, but is defunct, and needs to be recreated. The program
+	// no longer gets triggered at this point and the link needs to be removed
+	// to proceed.
+	case errors.Is(err, unix.ENOLINK):
+		if err := os.Remove(pin); err != nil {
+			return fmt.Errorf("unpinning defunct link %s: %w", pin, err)
+		}
+
+		log.Infof("Unpinned defunct link %s for program %s", pin, progName)
+
+		// Wrap in os.ErrNotExist so the caller needs to look for one error.
+		return fmt.Errorf("unpinned defunct link: %w", os.ErrNotExist)
+
+	// No existing link found, continue trying to create one.
+	case errors.Is(err, os.ErrNotExist):
+		log.Debugf("No existing link found at %s for program %s", pin, progName)
+		return err
+
+	case err != nil:
+		return fmt.Errorf("updating link %s for program %s: %w", pin, progName, err)
+	}
+
+	log.Infof("Updated link %s for program %s", pin, progName)
+	return nil
+}
+
+// detachNetkit attempts to open the link at progName in bpffsDir
+// and unpins it. Only returns unrecoverable errors. Returns nil if
+// the link doesn't exist or if removal was successful.
+func detachNetkit(bpffsDir, progName string) error {
+	pin := filepath.Join(bpffsDir, progName)
+	err := bpf.UnpinLink(pin)
+	if err == nil {
+		log.Infof("Removed netkit link at %s", pin)
+		return nil
+	}
+	if errors.Is(err, os.ErrNotExist) {
+		return nil
+	}
+
+	// The pinned link exists, something went wrong unpinning it.
+	return fmt.Errorf("unpinning netkit link: %w", err)
+}
+
+// hasCiliumNetkitLinks returns true if device has a Cilium-managed
+// netkit program with the given attach type.
+func hasCiliumNetkitLinks(device netlink.Link, attach ebpf.AttachType) (bool, error) {
+	if device.Type() != "netkit" {
+		// Not a netkit device, therefore also no netkit links.
+		return false, nil
+	}
+	result, err := link.QueryPrograms(link.QueryOptions{
+		Target: int(device.Attrs().Index),
+		Attach: attach,
+	})
+	if errors.Is(err, unix.EINVAL) {
+		// Attach type likely not supported, kernel doesn't support netkit.
+		return false, nil
+	}
+	if err != nil {
+		return false, fmt.Errorf("querying %s netkit programs for device %s: %w", attach, device.Attrs().Name, err)
+	}
+	if result == nil || len(result.Programs) == 0 {
+		return false, nil
+	}
+
+	for _, p := range result.Programs {
+		prog, err := ebpf.NewProgramFromID(p.ID)
+		if err != nil {
+			return false, fmt.Errorf("opening program with id %d: %w", p.ID, err)
+		}
+		defer prog.Close()
+
+		pi, err := prog.Info()
+		if err != nil {
+			continue
+		}
+		if strings.HasPrefix(pi.Name, "cil_") {
+			return true, nil
+		}
```

Comment:
Is it valid to have another netlink type in a pod we are attaching to? Maybe EEXISTs if its not "cil_"?

Reply:
Here we specifically probe for netkit peer type and we expect Cilium BPF progs which have cil_ prefix. If there is another prefix attached, we don't care. This is mainly used for watchdog to check if Cilium is attached everywhere.